Coreos kube-up now with less cloud init

euank opened the pull request:
This update includes significant refactoring. It moves almost all of the
logic into bash scripts, modeled after the `gci` cluster scripts.

The reason to do this is:
1. Avoid duplicating the saltbase manifests by reusing gci's parsing logic (easier maintenance)
2. Take an incremental step towards sharing more code between gci/trusty/coreos, again for better maintenance
3. Pave the way for making future changes (e.g. improved rkt support, kubelet support) easier to share

The primary differences from the gci scripts are the following:
1. Use of the `/opt/kubernetes` directory over `/home/kubernetes`
2. Support for rkt as a runtime
3. No use of logrotate
4. No use of `/etc/default/`
5. No logic related to noexec mounts or gci-specific firewall-stuff

It will make sense to move 2 over to gci, as well as perhaps a few other small improvements. That will be a separate PR for ease of review.

Ref #29720, this is a part of that because it removes a copy of them.

Fixes #24165

cc @yifan-gu 

Since this logic largely duplicates logic from the gci folder, it would be nice if someone closely familiar with that gave an OK or made sure I didn't fall into any gotchas related to that, so cc @andyzheng0831

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33965)

<!-- Reviewable:end -->

euank committed:
cluster/gce: copy gci -> coreos

This is for reviewing ease as the following commits introduce changes
to make the coreos kube-up deployment share significant code with the
gci code.

yifan-gu commented on cluster/gce/coreos/README.md:
```diff
@@ -0,0 +1,8 @@
+# CoreOS image
+
+The [CoreOS operating oystem](https://coreos.com/why/) is a Linux distribution optimized for running containers securely at scale.
```
s/oystem/system

yifan-gu commented:
@euank Looks like we don't need `cluster/gce/coreos/configure-kubelet.sh` as well.

yifan-gu commented on cluster/gce/coreos/configure-helper.sh:
```diff
@@ -0,0 +1,1147 @@
+function setup-kubelet-dir {
+    echo "Making /var/lib/kubelet executable for kubelet"
+    mount --bind /var/lib/kubelet /var/lib/kubelet/
+    mount -B -o remount,exec,suid,dev /var/lib/kubelet    
```
Some tailing spaces

yifan-gu commented on cluster/gce/coreos/configure.sh:
```diff
@@ -0,0 +1,173 @@
+#!/bin/bash
+
+# Copyright 2015 The Kubernetes Authors.
```
2016

yifan-gu commented on cluster/gce/coreos/configure.sh:
```diff
@@ -0,0 +1,173 @@
+    -o "${tmp_kube_env}" \
+    http://metadata.google.internal/computeMetadata/v1/instance/attributes/kube-env
+  # Convert the yaml format file into a shell-style file.
+  sed 's/: /=/' < "${tmp_kube_env}" > "${KUBE_DIR}/kube-env"
```
@euank Have you verified that this will always work for today's kube env yamls?

euank replied:
It does work as best as I can tell based on sourcing the file and printenv and based on skimming it.

Any specific concerns?

yifan-gu replied:
@euank Thank you.

yifan-gu commented:
LGTM other than the comments above. Thanks for the work!!! @euank

euank commented:
As a hint for reviewers, doing a (vim)diff between the `gci` and the `coreos` versions of `configure.sh` and `configure-helper.sh` is incredibly useful for review.

euank commented:
Addressed your comments and manually tested. Once I re-added the hostnamectl thing you had, a lot of the issues I had just went away.

I don't know of anything wrong with this now. My previous review tip of vim-diffing with gci still applies as a way to make reviewing this big diff much easier.

euank commented:
@s-urbaniak I don't know if you have any familiarity with the `cluster/gce/*` stuff, but if you'd be able to review it'd be appreciated!

euank commented:
cc @roberthbailey @mikedanese, since you two are listed as owners in `./cluster`. Any review you're willing/able to give is appreciated!

yifan-gu commented:
LGTM for the changes under `cluster/gce/coreos` dir

roberthbailey commented:
can you please squash your commits?

euank commented:
@roberthbailey I've squashed them a bit more, but I think the current split is the right level of separation between the different changes needed to get this to work.

euank commented:
@k8s-bot gci gce e2e test this

yifan-gu commented:
@robertabbott Do you mind take another look? Feel free to assign someone else if you don't have time.

euank commented:
@roberthbailey This is one of the rare cases where making smaller chunks might be harder to review actually.

Because this is largely based on the `cluster/gce/gci` code, diffing these files with those ends up being fairly readable.
I'd be happy to add a first commit prior to these that simply copies gci -> coreos and then apply these changes on top of that if you'd prefer that to diffing locally.

Unfortunately, these changes are fairly strongly intermingled

yifan-gu commented:
@roberthbailey 

I think more details on the PR might be helpful:

The 1st commit is mainly copying files from `configure.sh` and `configure-helper.sh` from `cluster/gce/gci` to `cluster/gce/coreos`, with very limited changes. https://github.com/kubernetes/kubernetes/pull/33965#issuecomment-251478825

The 4th commit is removing old files.
Other commits are just several line changes.


@euank 

One suggestion to make things easier for review, would be split this PR into 2.
- PR1: Copying `configure.sh` and `configure-helper.sh`, removing all old codes.
- PR2: Modify based on the previous PR1.

euank committed:
cluster/coreos: update veify-flags exceptions

euank committed:
cluster/coreos: update to gci based implementation

This update includes significant refactoring. It moves almost all of the
logic into bash scripts, modeled after the `gci` cluster scripts.

The primary differences between the two are the following:
1. Use of the `/opt/kubernetes` directory over `/home/kubernetes`
2. Support for rkt as a runtime
3. No use of logrotate
4. No use of `/etc/default/`
5. No logic related to noexec mounts or gci-specific firewall-stuff

euank committed:
cluster/coreos: delete mounter

We don't use this bit of gci currently.

euank commented:
Because you can't really make a PR on another PR on github, I just went straight for PR2.

@roberthbailey To make things easier to review, I have one commit that does `rm -rf cluster/gce/coreos; cp -R cluster/gce/gci cluster/gce/coreos`.

The diff between the directories is 0 after the first commit, and the next commit layers the changes needed by CoreOS (mainly removal of things CoreOS doesn't have or care about).

You can also manually test with e.g.:
```
export KUBE_SSH_USER=core
export KUBE_OS_DISTRIBUTION=coreos
export OS_DISTRIBUTION=coreos
export KUBE_GCE_MASTER_PROJECT=coreos-cloud
export KUBE_GCE_MASTER_IMAGE=coreos-alpha-1214-0-0-v20161028
export KUBE_GCE_NODE_PROJECT=${KUBE_GCE_MASTER_PROJECT}
export KUBE_GCE_NODE_IMAGE=${KUBE_GCE_MASTER_IMAGE}

./cluster/kube-up
```

euank commented:
@roberthbailey Does making the first commit a straight copy of the gci folder help for you review-wise? If not, have an idea for how this might be more reviewable?

roberthbailey commented:
@euank - yes, that helps. But I'm still pretty starved for review time and I'm ooo today so this may not get reviewed before the feature freeze on monday.

mikedanese commented:
@k8s-bot bazel test this, issue https://github.com/kubernetes/test-infra/issues/1152

euank commented:
Rebased out the conflicts @roberthbailey ...

mikedanese commented:
@k8s-bot bazel test this, issue kubernetes/test-infra#1152

euank commented:
Ping @roberthbailey 

As before, it's a clean "mv" followed by needed changes in the second commit, so the second commit is the important one to review. I managed to reduce it to a diff of "+287 −511" (mainly by not renaming KUBE_HOME -> KUBE_DIR even though the latter name fits a little better).

roberthbailey commented:
@euank -- sorry, i've been spending all of my review bandwidth working on 1.5 blockers. I'll circle back to this PR once those are finished (hopefully at the end of next week if we ship on time). I haven't forgotten about it.

euank commented:
If you have someone else you can suggest to review this, please do!

euank commented:
Ping @roberthbailey; as before, if you don't have spare cycles but can suggest someone else, that would be helpful as well.

euank commented:
Ping @roberthbailey

roberthbailey commented on build/common.sh:
```diff
@@ -649,6 +649,50 @@ function kube::build::start_rsyncd_container() {
+# such as GCI and Ubuntu Trusty. We directly copy manifests from
+# cluster/addons and cluster/saltbase/salt. The script of cluster initialization
+# will remove the salt configuration and evaluate the variables in the manifests.
+function kube::release::package_kube_manifests_tarball() {
```
Isn't this redundant with the function of the same name in build/lib/release.sh?

euank replied:
The was a rebase error. After fixing it, it's a 3-line diff in the original function

I'll update the PR after verifying things still work..

Thanks for catching it!

roberthbailey commented on cluster/gce/coreos/configure.sh:
```diff
@@ -14,30 +14,10 @@
 set -o nounset
 set -o pipefail
 
-function set-broken-motd {
```
FYI We've found that setting the motd is really useful for debugging node startup failure.

roberthbailey commented:
@euank 

Sorry for the long delay on the code review. I only two concerns:

1. It seems you've duplicated some code between shell scripts. They are likely to diverge in the future and it would be nice to consolidate them into one location.

1. Can you verify that coreos/gci are mostly copies of each other now rather than intimately sharing the node startup scripts (as appears to be the case from https://github.com/euank/kubernetes/tree/3536ecc6ff7f043d5a104b92b803881e5f7e33af/cluster/gce/coreos)? I'm concerned that we'll make a change to gci that will break the coreos startup scripts and won't notice as we don't have continuous testing of them.

bgrant0607 commented:
@grodrigues3 Is there a PR that adds an OWNERS file for cluster/gce/coreos? Could we break it out and get it in earlier than the rest?

roberthbailey commented:
@bgrant0607 -- juju and vagrant are the only directories under cluster that have an OWNERS file.

euank commented:
@roberthbailey Per the original comment, number 2. This PR is a first step towards "an incremental step towards sharing more code between gci/trusty/coreos, again for better maintenance"

I want to do that in a followup PR because that change will also be intrusive to GCI.

I don't understand what you mean by "sharing startup scripts". gci/coreos folders do not share code yet (though I want them to in the future as above).
I don't think any change in the gci folder will impact the coreos folder at this time.

roberthbailey commented:
@euank - thanks. That answered by second question. My only remaining question is about the duplication of functions in shell.

bgrant0607 commented:
@roberthbailey I was suggesting that we create a new OWNERS file

roberthbailey commented:
Yup. I asked @euank to send a quick PR to add one.

euank commented:
@roberthbailey I addressed your comment and rebased it back on master again to pick up any missing gci changes.

I validated that it works for me with my kube-up environment of:
```
export KUBE_SSH_USER=core
export KUBE_OS_DISTRIBUTION=coreos

export KUBE_GCE_MASTER_PROJECT=coreos-cloud
export KUBE_GCE_MASTER_IMAGE=$(gcloud compute images list --project coreos-cloud --filter  coreos-alpha --format=json | jq '.[0].name' -r)
export KUBE_GCE_NODE_PROJECT=coreos-cloud
export KUBE_GCE_NODE_IMAGE=${KUBE_GCE_MASTER_IMAGE}
```

I'll send an owners pr pronto as well including myself, @yifan-gu, and @ethernetdan as owners.

roberthbailey commented:
@euank looks like it needs another rebase.

euank commented:
Rebased ..

roberthbailey commented:
/lgtm
